src/librustc_trans/back/link.rs

```diff
@@ -1184,20 +1227,50 @@ fn add_upstream_rust_crates(cmd: &mut Linker,
     }
 
     // Same thing as above, but for dynamic crates instead of static crates.
-    fn add_dynamic_crate(cmd: &mut Linker, sess: &Session, cratepath: &Path) {
+    fn add_dynamic_crate(cmd: &mut Linker, sess: &Session, tmpdir: &Path, cratepath: &Path) {
         // If we're performing LTO, then it should have been previously required
         // that all upstream rust dependencies were available in an rlib format.
         assert!(!sess.lto());
 
-        // Just need to tell the linker about where the library lives and
-        // what its name is
-        let parent = cratepath.parent();
-        if let Some(dir) = parent {
-            cmd.include_path(&fix_windows_verbatim_for_gcc(dir));
-        }
         let filestem = cratepath.file_stem().unwrap().to_str().unwrap();
-        cmd.link_rust_dylib(&unlib(&sess.target, filestem),
-                            parent.unwrap_or(Path::new("")));
+        if sess.target.target.options.is_like_msvc {
+            let ref src = cratepath.with_extension("dll.rlib");
+            let ref dst_dir = tmpdir.join(filestem);
+            let _ = fs::create_dir_all(dst_dir);
+            let ref dst = dst_dir.join(filestem).with_extension("dll.lib");
+
+            // Modify import library to remove metadata before linking
```

Comment:
Would it be possible to emit the symbol with a visibility level that we don't have to do this?